aiida_raspa/workchains/base.py

```diff
@@ -45,18 +45,162 @@ def report_error_handled(self, calculation, action):
-        """
+        :param action: a string message with the action taken"""
+
         arguments = [calculation.process_label, calculation.pk, calculation.exit_status, calculation.exit_message]
         self.report('{}<{}> failed with exit status {}: {}'.format(*arguments))
         self.report('Action taken: {}'.format(action))
 
-
-@register_error_handler(RaspaBaseWorkChain, 570)
-def _handle_timeout(self, calculation):
-    """Error handler that restarts calculation finished with TIMEOUT ExitCode."""
-    if calculation.exit_status == RaspaCalculation.spec().exit_codes.TIMEOUT.status:
+    @process_handler(priority=570, exit_codes=RaspaCalculation.exit_codes.TIMEOUT, enabled=True)
+    def handle_timeout(self, calculation):
+        """Error handler that restarts calculation finished with TIMEOUT ExitCode."""
         self.report_error_handled(calculation, "Timeout handler. Adding remote folder as input to use binary restart.")
         self.ctx.inputs.parent_folder = calculation.outputs.remote_folder
+        return ProcessHandlerReport(False)
+
+    @process_handler(priority=400, enabled=False)
+    def check_widom_convergence(self, calculation):
+        """Checks whether a Widom particle insertion is converged. The check is based on the
+        error bar of the Henry coefficient."""
+
+        conv_threshold = 0.1
+        additional_cycle = 0
+
+        output_widom = calculation.outputs.output_parameters.get_dict()
+        structure_label = list(calculation.get_incoming().nested()['framework'].keys())[0]
+        conv_stat = []
+
+        for comp in calculation.inputs.parameters['Component']:
+            kh_average_comp = output_widom[structure_label]["components"][comp]["henry_coefficient_average"]
+            kh_dev_comp = output_widom[structure_label]["components"][comp]["henry_coefficient_dev"]
+
+            error = round((kh_dev_comp / kh_average_comp), 2)
+            if error <= conv_threshold:
+                conv_stat.append(True)
+            else:
+                conv_stat.append(False)
+
+        if not all(conv_stat):
+
+            self.report("Widom particle insertion calculationulation is NOT converged: repeating with more trials...")
+            self.ctx.inputs.retrieved_parent_folder = calculation.outputs['retrieved']
+            self.ctx.inputs.parameters = modify_number_of_cycles(self.ctx.inputs.parameters,
+                                                                 additional_init_cycle=Int(0),
+                                                                 additional_prod_cycle=Int(additional_cycle))
+            return ProcessHandlerReport(False)
+
+        return None
+
+    @process_handler(priority=410, enabled=False)
+    def check_gcmc_convergence(self, calc):
+        """Checks whether a GCMC calc is converged. Checking is based on the error bar on average loading."""
+        conv_threshold = 0.1
+        additional_init_cycle = 0
+        additional_prod_cycle = 0
```

Comment:
Here, it needs to change to accept `conv_threshold` and `additional_init_cycle` and `additiona_prod_cycle` as arguments.